javaslang/src/main/java/javaslang/collection/Stream.java

```diff
@@ -267,7 +267,11 @@ public boolean hasNext() {
 
             @Override
             public T next() {
-                return elements[i++];
+                if(hasNext()) {
+                    return elements[i++];
+                } else {
+                    throw new NoSuchElementException();
+                }
```

Comment:
This is a valid optimization. Please revert it according to my [comment](https://github.com/javaslang/javaslang/pull/1228#issuecomment-201316395).